### PR TITLE
#311 Tooltip enhancements

### PR DESCRIPTION
**Pull request by patric-eberle**
## Pull request
This PR improves the tooltip implementation according to the findings from the last project. It will mainly prevent duplicated items in the code. I also decided to change a related TS `interface` to `type`, which had some side effects.

Note: there is still a bug when using multiple tooltips beside each other. Since this issue already exists in the source project, and I was a bit overwhelmed by the issue, I made a bug ticket for now: #337 
 
### Ticket
#311 
 
### Browser testing
- http://localhost:5173/styleguide/sandbox/tooltips
 
### Checklist
- [x] I merged the current development branch (before testing)
- [x] Added JSDoc and styleguide demo
- [x] Tested all links in project relevant browsers
- [x] Tested all links in different screen sizes
- [x] Did run automated tests and linters
- [x] Did assign ticket
- [x] Double checked target branch
 
## Review/Test checklist
- [ ] Did review code and documentation
- [ ] Tested all links in project relevant browsers
- [ ] Tested all links in different screen sizes
- [ ] Did check accessibility (Wave, only errors)
- [ ] Re-assign ticket to developer

### CODE REVIEW

**Comment by patric-eberle** [comment]
@BeneRichi can you please also review my changes on this branch? THX.

File: src/plugins/tooltip/c-tooltip.vue [hunk 1]
```diff
@@ -2,24 +2,29 @@
...
-    <Teleport to="body">
-      <div
-        ref="tooltip"
-        :class="b('tooltip-wrapper')"
-        @load.capture="onLoad"
-      >
-        <div :class="b('tooltip')">
-          <!-- @slot This slot will be used for the tooltip content. -->
-          <slot name="tooltip"></slot>
+    <Teleport v-if="tooltipInitialized" to="body">
```
Comment (BeneRichi) [inline src/plugins/tooltip/c-tooltip.vue:12]: Is using `v-if` directly on the Teleport the right approach? In a Vue 3 docs example they are applying it to the child container inside the Teleport.

https://vuejs.org/guide/built-ins/teleport.html

File: src/plugins/tooltip/c-tooltip.vue [hunk 2]
```diff
@@ -33,9 +38,7 @@
   } from 'vue';
   import { createPopper, Instance, Options } from '@popperjs/core';
```
Comment (BeneRichi) [inline src/plugins/tooltip/c-tooltip.vue:39]: suggestion: In my recent Vue3 upgrade, we switched from `popper.js` to `floating-vue`. Not sure if it's project-specific, but it is maybe worth to discuss it.

File: src/plugins/tooltip/c-tooltip.vue [hunk 3]
```diff
@@ -44,23 +47,16 @@
   }
 
   interface Data {
```
Comment (BeneRichi) [inline src/plugins/tooltip/c-tooltip.vue:49]: todo: Please use `type`

Comment (BeneRichi) [reply src/plugins/tooltip/c-tooltip.vue:49]: On line45 for the `Setup` as well.

File: src/plugins/tooltip/directives/directive.ts [hunk 4]
```diff
@@ -11,41 +11,80 @@ import {
...
 
 const storageKey = Symbol('Tooltip directive instance');
+const tooltipAnchor = Symbol('The current tooltip anchor');
 
 type TooltipEvent = {
   [key: string]: EventListener;
 }
 
-interface TooltipElement extends HTMLElement {
+interface AnchorElement extends HTMLElement {
```
Comment (BeneRichi) [inline src/plugins/tooltip/directives/directive.ts:23]: todo: Please make the necessary updates:

`type AnchorElement = HTMLElement & {
  [storageKey]: {
    isHidden: boolean;
    popper: Instance;
    events: TooltipEvent[];
    content: string;
  };
}
`

File: src/plugins/tooltip/directives/directive.ts [hunk 5]
```diff
@@ -11,41 +11,80 @@ import {
...
   [storageKey]: {
     isHidden: boolean;
     popper: Instance;
     events: TooltipEvent[];
+    content: string;
   };
 }
 
+interface Tooltip extends HTMLDivElement {
```
Comment (BeneRichi) [inline src/plugins/tooltip/directives/directive.ts:32]: todo: Please make the necessary updates:

`type Tooltip = HTMLDivElement & {
  [tooltipAnchor]?: HTMLElement;
}`

File: src/styleguide/routes/r-tooltips.vue [hunk 6]
```diff
@@ -6,9 +6,11 @@
         Disable/Enable Tooltip
       </h2>
 
-      <div :class="b('link')">
-        <span v-tooltip:[position].mouseover="'Content for the tooltip'">Hover me to see the tooltip.</span>
-      </div>
+      <ul>
+        <li :class="b('link')">
```
Comment (BeneRichi) [inline src/styleguide/routes/r-tooltips.vue:10]: suggestion: I feel like `link` might not be the best name for the `<li>` element showcasing the tooltip. How about `item` or `element` for better clarity in this context?

**Review by BeneRichi — CHANGES_REQUESTED** [review]
Browser review looks fine 👍

File: src/plugins/tooltip/c-tooltip.vue (hunk 2 above)
Comment (patric-eberle) [reply src/plugins/tooltip/c-tooltip.vue:39]: @BeneRichi good input. Can you create an additional ticket to replace it with information where to find the related code. Not sure where you added this.

File: src/plugins/tooltip/c-tooltip.vue [hunk 7]
```diff
@@ -172,50 +165,30 @@
...
-      },
-
       /**
        * Refreshes the popper if images did load inside of it.
        */
-      onLoad(): void {
+      updatePopperInstance(): void {
         this.popperInstance?.update();
       },
     },
```
[the remark this answers is not in the document]
Comment (patric-eberle) [reply src/plugins/tooltip/c-tooltip.vue:194]: @BeneRichi the file is located in a subfolder in relation to the importing file. The `./` at the beginning of the URL tells the browser, that the URL starts from the currents file URL. Thus: "import `styles.scss` from the subfolder `styles` located in the same folder as this file."

**Comment by patric-eberle** [comment]
@BeneRichi applied the requested changes. Please also check my responses to some of the comments and check code again.

File: src/plugins/tooltip/c-tooltip.vue (hunk 2 above)
Comment (BeneRichi) [reply src/plugins/tooltip/c-tooltip.vue:40]: I have opened an issue: https://github.com/valantic/vue-template/issues/354.

**Review by BeneRichi — APPROVED** [review]
I've tested again. Works still fine. Amazing!

**Pull request merged**